### DISCUSSION TITLE
Any plans on supporting `QSerialPort` as first party socket?

### DISCUSSION THREAD

**Discussion by ptsouchlos** [danvratil/qcoro#102] [Q&A] [answered]
I noticed that some subclasses of `QIODevice` are supported such as `QLocalSocket`. Are there any plans to support `QSerialSocket` directly?

**Comment by danvratil** [comment]
I don't see a reason why it wouldn't be possible to support it. Probably the hardest thing will be creating a test infrastructure for it :-)

**Reply by danvratil** [reply]
Tracking issue: https://github.com/danvratil/qcoro/issues/104

**Reply by ptsouchlos** [reply]
Fantastic, thank you!

As for testing, you could probably mock an "echo" device right? I'd be willing to help with this as well.

**Reply by danvratil** [reply]
It would probably be similar to the TestHttpServer, so not just echo, but also sending a big response in chunks and simulating a delay.

Looks like I can just use pseudoterminals on Linux for this: https://stackoverflow.com/questions/49544318/using-linux-pseudoterminal-to-test-qserialport/49547001#49547001, which is good enough for me.

**Comment by danvratil** [comment, marked as answer]
Just looking at the code of `QSerialPort` (which I assume is what you meant instead of `QSerialSocket`) and I think it should work out-of-the-box with the existing `QCoroIODevice` wrapper.

The reason the other sockets got their own wrappers is that they have asynchronous state which needs to be taken into account. But `QSerialPort` doesn't have async state and `open()` and `close()` are synchronous operations, so I think it shoud work just fine with `QCoroIODevice`.

Please try out and let me know if there are any issues. If it turns out there's a need for a dedicated wrapper type, I'm totally fine with implementing it.

I realize that it's confusing why some `QIODevice`s need a special wrapper and some work just with `QCoroIODevice` and it's hard to know whether the type just works or whether it needs a wrapper to be implemented. I'll try to make it clearer in the documentation of `QCoroIODevice`.

**Reply by ptsouchlos** [reply]
Sounds good, this is what I thought at first but I figured I would ask anyway in case there was something I missed. Thanks for taking a look! I think updating the docs would be nice and maybe still adding tests for the `QSerialPort` might be good? Although might be overkill since you already have tests for `QIODevice`.
